bika/lims/browser/reports/productivity_analysesperclient.py

```diff
@@ -20,170 +26,167 @@ class Report(BrowserView):
     template = ViewPageTemplateFile("templates/report_out.pt")
 
     def __init__(self, context, request, report=None):
-        self.report = report
         BrowserView.__init__(self, context, request)
+        self.report = report
+        self.headings = {
+            'header': _("Analysis requests and analyses per client"),
+            'subheader': _("Number of Analysis requests and analyses per client"),
+        }
+        self.formats = {
+            'columns': 3,
+            'col_heads': [
+                _('Client'),
+                _('Number of requests'),
+                _('Number of analyses')],
+            'class': ''
+        }
 
     def __call__(self):
-        # get all the data into datalines
+        parms = []
 
-        pc = getToolByName(self.context, 'portal_catalog')
-        bac = getToolByName(self.context, 'bika_analysis_catalog')
-        bc = getToolByName(self.context, 'bika_catalog')
-        rc = getToolByName(self.context, 'reference_catalog')
+        # Base query
+        query = dict(portal_type="Analysis", sort_on='getClientTitle')
+
+        # Filter by client
+        self.add_filter_by_client(query, parms)
+
+        # Filter by date
+        self.add_filter_by_date(query, parms)
+
+        # Filter analyses by review_state
+        self.add_filter_by_review_state(query, parms)
+
+        # Filter analyses by cancellation state
+        self.add_filter_by_cancellation_state(query, parms)
+
+        # Filter analyses by worksheet state
+        self.add_filter_by_worksheet_state(query, parms)
```

Comment:
All of the methods change the state of `query` and `params` indirectly. I would rather add methods which return a query fragment which can be updated to the original query like so:
`query.update(self.get_review_state_query())`

Reply:
Yes, you're right. Will change this

Reply:
Ummm... then I'd need to do the update for both `query` and `parms`. I think the whole productivity reporting machinery needs to be rebuilt/refactored and all these functions will probably be left behind then. Thinking about `senaite.databox`